### PR TITLE
Implement std::error::Error and document error variants

### PR DESCRIPTION
**Pull request by curiousleo**
Rust errors should implement `std::error::Error` for interoperability.
https://rust-lang.github.io/api-guidelines/interoperability.html#c-good-err

In addition, public modules and definitions should be documented. Since I was looking at this module anyway, I made a start here.
https://rust-lang.github.io/api-guidelines/documentation.html

**Before:**
Screenshot_2020-02-10 nixrust Error - Rust(1)

**After:**
Screenshot_2020-02-10 nixrust Error - Rust

### CODE REVIEW

**Comment by edolstra** [comment]
This conflicts with some refactoring to `Error` that I've made (but haven't finished yet) to split it into multiple `Error` types (and maybe use error_chain or failure, but I haven't decided on that yet).

File: nix-rust/src/error.rs [hunk 1]
```diff
@@ -1,30 +1,55 @@
+//! Error types
+
 use std::fmt;
 
+/// An error that can occur in this crate.
 #[derive(Debug)]
 pub enum Error {
+    /// The store path could not be resolved.
```
Comment (edolstra) [inline nix-rust/src/error.rs]: IMHO this isn't really desirable since it duplicates the error descriptions in the `fmt::Display` trait implementation.

Comment (curiousleo) [reply nix-rust/src/error.rs]: I see what you mean. The error variants have very explicit names too, so perhaps these docs don't add much.

**Comment by curiousleo** [comment]
> This conflicts with some refactoring to `Error` that I've made (but haven't finished yet) to split it into multiple `Error` types (and maybe use error_chain or failure, but I haven't decided on that yet).

Okay, good to know! I'll close this PR then.

**Comment by Ekleog** [comment]
@edolstra FWIW, error-chain is long dead, and failure has been halfway merged into the stdlib -- the part that wasn't is now available in `thiserror`, which is as a consequence the crate new projects should probably go to by default.

**Pull request closed without merging**